Add support for compile-only dependencies and class path expansion (properties, globs)

alexarchambault opened the pull request:
I've been needing these features in some of the things I'm working on, so I quickly hacked that

alexarchambault committed:
Add support for compile-only dependencies

alexarchambault committed:
Remove spurious character in error message

alexarchambault committed:
Expand class path arguments with class-path-util

So that properties and some globs are expanded. This makes things like
```
//> using jars "${spark.home}/jars/*:${spark.dist.classpath}"
```
work for example (provided users set the `spark.home` and
`spark.dist.classpath` properties)

alexarchambault commented:
I guess the last two commits need a bit of doc…

Gedochao commented on modules/options/src/main/scala/scala/build/Artifacts.scala:
```diff
@@ -356,14 +396,19 @@ object Artifacts {
         .map(_.flatten)
     }
 
+    coursier.Artifacts()
```
?

Gedochao commented on modules/options/src/main/scala/scala/build/Artifacts.scala:
```diff
@@ -415,24 +483,31 @@ object Artifacts {
+    res
+  }
+
+  private def fetchKeepFetcher(
```
This name is a bit confusing, I think... some scaladocs would help, maybe 🤔 
It's hard to get around between `fetch`, `fetch0` and `fetchKeepFetcher`...

Gedochao commented on modules/cli/src/main/scala/scala/cli/exportCmd/MillProjectDescriptor.scala:
we might want some `export` test for compile-only dependencies, for both Mill & SBT

alexarchambault committed:
Split --extra-jar and -classpath options handling

So that the handling of --extra-jar can be refined in subsequent commits,
while we can retain some legacy features in -classpath

Gedochao commented:
@alexarchambault What's the current status of this PR? Any plans to finish it in the near future?

alexarchambault commented:
> What's the current status of this PR? Any plans to finish it in the near future?

I do, likely in the coming weeks, I'd say

lwronski commented:
> > What's the current status of this PR? Any plans to finish it in the near future?
> 
> I do, likely in the coming weeks, I'd say

@alexarchambault  Hi Alex, I hope you're doing well. I'm just checking to see if you have time to finish this PR soon. If you're busy, I can help to finish it, especially to add support for the compile-only dependencies. Let me know your preference.

Gedochao commented:
Closing this, as it seems to have been largely delivered already by @lwronski in #2299 (#817).
The remaining requirements are already tracked separately under #2027.
